kmyth_log/src/kmyth_log.c

```diff
@@ -0,0 +1,350 @@
+void set_applog_severity_threshold(int new_severity_threshold)
+{
+  if ((new_severity_threshold >= 0) && (new_severity_threshold <= 7))
+  {
+    log_settings.applog_severity_threshold = new_severity_threshold;
+  }
+  else
+  {
+    // do nothing if invalid, but warn user
+    fprintf(stderr, "set_applog_severity_threshold(): ");
+    fprintf(stderr, "input (%d) invalid - unchanged ", new_severity_threshold);
+    fprintf(stderr, "(%d)\n", log_settings.applog_severity_threshold);
+  }
+}
+
+//############################################################################
+// set_syslog_facility()
+//   - valid values: (0<<3) - ((LOG_NFACILITIES-1)<<3)
+//############################################################################
+void set_syslog_facility(int new_syslog_facility)
+{
+  if ((LOG_FAC(new_syslog_facility) >= 0) &&
+      (LOG_FAC(new_syslog_facility) <= (LOG_NFACILITIES - 1)))
+  {
+    log_settings.syslog_facility = new_syslog_facility;
+  }
+  else
+  {
+    // do nothing if invalid, but warn user
+    fprintf(stderr, "set_syslog_facility(): ");
+    fprintf(stderr, "input (%d) ", LOG_FAC(new_syslog_facility));
+    fprintf(stderr, "invalid - unchanged ");
+    fprintf(stderr, "(%d)\n", LOG_FAC(log_settings.syslog_facility));
+  }
+}
+
+//############################################################################
+// set_syslog_severity_threshold()
+//   - valid values: 0-7
+//############################################################################
+void set_syslog_severity_threshold(int new_severity_threshold)
+{
+  if ((new_severity_threshold >= 0) && (new_severity_threshold <= 7))
+  {
+    log_settings.syslog_severity_threshold = new_severity_threshold;
+  }
+  else
+  {
+    // do nothing if invalid, but warn user
+    fprintf(stderr, "set_syslog_severity_threshold(): ");
+    fprintf(stderr, "input (%d) invalid - unchanged ", new_severity_threshold);
+    fprintf(stderr, "(%d)\n", log_settings.syslog_severity_threshold);
+  }
+}
+
+//############################################################################
+// get_severity_str()
+//############################################################################
+void get_severity_str(int severity_val_in, char **severity_str_out)
+{
+  switch (severity_val_in)
+  {
+  case LOG_EMERG:
+    asprintf(severity_str_out, "EMERGENCY");
+    break;
+  case LOG_ALERT:
+    asprintf(severity_str_out, "ALERT");
+    break;
+  case LOG_CRIT:
+    asprintf(severity_str_out, "CRITICAL");
+    break;
+  case LOG_ERR:
+    asprintf(severity_str_out, "ERROR");
+    break;
+  case LOG_WARNING:
+    asprintf(severity_str_out, "WARNING");
+    break;
+  case LOG_NOTICE:
+    asprintf(severity_str_out, "NOTICE");
+    break;
+  case LOG_INFO:
+    asprintf(severity_str_out, "INFO");
+    break;
+  case LOG_DEBUG:
+    asprintf(severity_str_out, "DEBUG");
+    break;
+  default:
+    asprintf(severity_str_out, "CUSTOM(%d)", severity_val_in);
```

Comment:
How can a custom level be defined? Looking at log_even I guess using negative severity_val_in?

Reply:
A "custom" level greater than 7 would be inconsistent with syslog defined levels and would never satisfy the severity threshold. I guess a negative severity_val_in would meet the threshold, but would the inconsistent with the syslog defined levels and I don't know what value they might add.

In the log_event() function, therefore, I added a line that passes the input severity parameter to LOG_PRI() before using it. This will mask out all but the lowest three bits, forcing the value to the range 0-7. The default case in this switch statement, therefore, should actually never occur. Just in case, however, I changed the severity string result to "INVALID SEVERITY (%d)".